### PR TITLE
Some final touches for variadic types support

### PR DESCRIPTION
**Pull request by ilevkivskyi**
I decided to go again over various parts of variadic types implementation to double-check nothing is missing, checked interaction with various "advanced" features (dataclasses, protocols, self-types, match statement, etc.), added some more tests (including incremental), and `grep`ed for potentially unhandled cases (and did found few crashes). This mostly touches only variadic types but one thing goes beyond, the fix for self-types upper bound, I think it is correct and should be safe.

If there are no objections, next PR will flip the switch.

### CODE REVIEW

**Review by hauntsaninja — APPROVED** [review]
Ah, just saw I didn't do merge right, thanks for fixing! Tests and changes I understand look good

**Comment by ilevkivskyi** [comment]
@jhance @JukkaL To speed things up I am going to merge this later today, if there are no more comments.

**Review by JukkaL — APPROVED** [review]
Did a quick review pass, looks good.

**Pull request merged**
